<a href="https://colab.research.google.com/github/krish-navulla/SecureTLForIrisBiometrics/blob/main/Secure_IRIS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Google Drive Mounting and Git cloning

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
!git clone https://github.com/krish-navulla/SecureTLForIrisBiometrics.git


Cloning into 'SecureTLForIrisBiometrics'...
remote: Enumerating objects: 162, done.
remote: Counting objects: 100% (162/162), done.
remote: Compressing objects: 100% (104/104), done.
remote: Total 162 (delta 98), reused 97 (delta 56), pack-reused 0
Receiving objects: 100% (162/162), 234.52 KiB | 4.34 MiB/s, done.
Resolving deltas: 100% (98/98), done.


In [4]:
!rm -rf SecureTLForIrisBiometrics

In [3]:
%cd SecureTLForIrisBiometrics/


/content/SecureTLForIrisBiometrics


In [10]:
!git pull


remote: Enumerating objects: 7, done.
remote: Counting objects: 100% (7/7), done.
remote: Compressing objects: 100% (1/1), done.
remote: Total 4 (delta 3), reused 4 (delta 3), pack-reused 0
Unpacking objects: 100% (4/4), 386 bytes | 386.00 KiB/s, done.
From https://github.com/krish-navulla/SecureTLForIrisBiometrics
   fc93c92..3672b61  main       -> origin/main
Updating fc93c92..3672b61
Fast-forward
 dataset.py | 4 ++--
 models.py  | 4 +++-
 2 files changed, 5 insertions(+), 3 deletions(-)


In [18]:
!git add .


In [20]:
!git config --global user.email "navullavamshi2423@gmail.com"
!git config --global user.name "krish-navulla"

In [25]:
!git config --global credential.helper manager


In [21]:
!git commit -m "Vit_b_16_changes"

[main 6a2a042] Vit_b_16_changes
 10 files changed, 5 insertions(+), 5 deletions(-)
 create mode 100644 Securetl_Iris_Attention_model_16b_50.pth
 create mode 100644 Securetl_Iris_Attention_model_16b_50_trainhist.pk
 create mode 100644 Securetl_Iris_Attention_model_16b_50_validhist.pk
 create mode 100644 __pycache__/dataset.cpython-310.pyc
 create mode 100644 __pycache__/eval.cpython-310.pyc
 create mode 100644 __pycache__/losses.cpython-310.pyc
 create mode 100644 __pycache__/models.cpython-310.pyc
 create mode 100644 __pycache__/trainer.cpython-310.pyc


In [26]:
!git push

git: 'credential-manager' is not a git command. See 'git --help'.
fatal: could not read Username for 'https://github.com': No such device or address


In [27]:
!git remote -v


origin	https://github.com/krish-navulla/SecureTLForIrisBiometrics.git (fetch)
origin	https://github.com/krish-navulla/SecureTLForIrisBiometrics.git (push)


# Requirements

In [ ]:
!pip install torch
import torch

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
!pip install torchvision==0.*

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
!pip install certifi==2020.12.5 chardet==3.0.4 cycler==0.10.0 entropy-estimators==0.0.1 facenet-pytorch==2.3.0 idna==2.10 joblib==0.17.0 kiwisolver==1.3.1 matplotlib==2.2.2 numpy==1.19.4 Pillow==8.0.1 pyparsing==2.4.7 python-dateutil==2.8.1 pytz==2020.4 requests==2.25.0 scikit-learn==0.22.2.post1 scipy==1.5.4 six==1.15.0 torch==1.4.0 torchvision==0.5.0 urllib3==1.26.2


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 147.5/147.5 kB 4.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 133.4/133.4 kB 12.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 35.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.8/58.8 kB 2.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 301.5/301.5 kB 24.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 53.1/53.1 kB 5.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 37.3/37.3 MB 11.3 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.3/7.3 MB 90.7 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
 

In [ ]:
!pip install torchvision==0.5.0


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
ERROR: Could not find a version that satisfies the requirement torchvision==0.5.0 (from versions: 0.1.6, 0.1.7, 0.1.8, 0.1.9, 0.2.0, 0.2.1, 0.2.2, 0.2.2.post2, 0.2.2.post3, 0.12.0, 0.13.0, 0.13.1, 0.14.0, 0.14.1, 0.15.0, 0.15.1)
ERROR: No matching distribution found for torchvision==0.5.0


In [ ]:
!pip install facenet_pytorch

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 22.3 MB/s eta 0:00:00


In [ ]:
import facenet_pytorch
from facenet_pytorch.models.inception_resnet_v1 import InceptionResnetV1


In [ ]:
!pip show facenet_pytorch

Name: facenet-pytorch
Version: 2.5.3
Summary: Pretrained Pytorch face detection and recognition models
Home-page: https://github.com/timesler/facenet-pytorch
Author: Tim Esler
Author-email: tim.esler@gmail.com
License: 
Location: /usr/local/lib/python3.10/dist-packages
Requires: numpy, pillow, requests, torchvision
Required-by: 


In [ ]:
import os


In [5]:
!pip install entropy-estimators==0.0.1
!pip install facenet-pytorch

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Preparing metadata (setup.py) ... done
  Created wheel for entropy-estimators: filename=entropy_estimators-0.0.1-py3-none-any.whl size=8299 sha256=51579784ccc4b29f58095cd361b70c2273be84d33add20bc1ad2bee8e7df7653
  Stored in directory: /root/.cache/pip/wheels/0c/f6/6c/7730a2fad2f91b0b02dafb82e42601d820482c66fb9d9046c7
Successfully built entropy-estimators
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 18.3 MB/s eta 0:00:00


# Vision Transformer - Vit_b_32

In [5]:
%cd /content/SecureTL/


/content/SecureTL


In [ ]:
import numpy as np
import cv2
import torchvision
import torch

Mounted at /content/drive


Preparing Iris images for Training

In [32]:
!python face_prepare_ytfdb.py

Completed train triplets for identity no. 1
Completed train triplets for identity no. 2
Completed train triplets for identity no. 3
Completed train triplets for identity no. 4
Completed train triplets for identity no. 5
Completed train triplets for identity no. 6
Completed train triplets for identity no. 7
Completed train triplets for identity no. 8
Completed train triplets for identity no. 9
Completed train triplets for identity no. 10
Completed train triplets for identity no. 11
Completed train triplets for identity no. 12
Completed train triplets for identity no. 13
Completed train triplets for identity no. 14
Completed train triplets for identity no. 15
Completed train triplets for identity no. 16
Completed train triplets for identity no. 17
Completed train triplets for identity no. 18
Completed train triplets for identity no. 19
Completed train triplets for identity no. 20
Completed train triplets for identity no. 21
Completed train triplets for identity no. 22
Completed train tri

Iris Attention Model Training

In [34]:
!python face_train_securetl_model.py

Training model: Securetl_Iris_Attention_model_50
Epoch 1
........mini-batch no. 12 loss: 1.1390
....train loss: 1.0146 :: EER 0.2606 :: Canc_EER 0.4989 :: D_sys 0.5015
....valid loss: 1.0256 :: EER 0.3298 :: Canc_EER 0.5000 :: D_sys 0.3893
....Saving...
Epoch 2
........mini-batch no. 12 loss: 1.1085
....train loss: 1.0131 :: EER 0.2544 :: Canc_EER 0.4994 :: D_sys 0.5331
....valid loss: 1.0230 :: EER 0.3245 :: Canc_EER 0.5071 :: D_sys 0.4155
....Saving...
Epoch 3
........mini-batch no. 12 loss: 1.0591
....train loss: 1.0120 :: EER 0.2356 :: Canc_EER 0.4973 :: D_sys 0.5586
....valid loss: 1.0223 :: EER 0.2979 :: Canc_EER 0.5053 :: D_sys 0.4468
....Saving...
Epoch 4
........mini-batch no. 12 loss: 1.0844
....train loss: 1.0099 :: EER 0.2473 :: Canc_EER 0.4947 :: D_sys 0.5840
....valid loss: 1.0223 :: EER 0.2979 :: Canc_EER 0.4894 :: D_sys 0.4452
Epoch 5
........mini-batch no. 12 loss: 1.0830
....train loss: 1.0087 :: EER 0.2327 :: Canc_EER 0.4936 :: D_sys 0.6019
....valid loss: 1.0216 :: 

Testing Iris Attention Model

In [35]:
!python face_test_secure_model.py

Testing model: Securetl_Iris_Attention_model_50
EER 0.1833 at threshold 0.2185 :: Canc_EER 0.4833 at threshold 0.1367 :: d_sys 0.7767
Privacy Leakage Rate 1.0000, Secrecy Leakage 1.3290


Summary



1.   Epochs = 5; EER 0.2000 at threshold 0.1701 :: Canc_EER 0.5000 at threshold 0.1211 :: d_sys 0.6949
Privacy Leakage Rate 1.0000, Secrecy Leakage 1.3290
2.   Epochs = 50; EER 0.1833 at threshold 0.2185 :: Canc_EER 0.4833 at threshold 0.1367 :: d_sys 0.7767
Privacy Leakage Rate 1.0000, Secrecy Leakage 1.3290



# Inception Resnet model


preprocessing and triplet generation

In [39]:
!python iris_prepare_AMF_inception.py

Completed train triplets for identity no. 1
Completed train triplets for identity no. 2
Completed train triplets for identity no. 3
Completed train triplets for identity no. 4
Completed train triplets for identity no. 5
Completed train triplets for identity no. 6
Completed train triplets for identity no. 7
Completed train triplets for identity no. 8
Completed train triplets for identity no. 9
Completed train triplets for identity no. 10
Completed train triplets for identity no. 11
Completed train triplets for identity no. 12
Completed train triplets for identity no. 13
Completed train triplets for identity no. 14
Completed train triplets for identity no. 15
Completed train triplets for identity no. 16
Completed train triplets for identity no. 17
Completed train triplets for identity no. 18
Completed train triplets for identity no. 19
Completed train triplets for identity no. 20
Completed train triplets for identity no. 21
Completed train triplets for identity no. 22
Completed train tri

Training the Inception Resnet model

In [42]:
!python iris_train_inception_model.py

Training model: Securetl_Iris_Inception_model_50
Epoch 1
........mini-batch no. 12 loss: 1.0220
....train loss: 1.0042 :: EER 0.3393 :: Canc_EER 0.4752 :: D_sys 0.2909
....valid loss: 1.0062 :: EER 0.3989 :: Canc_EER 0.4924 :: D_sys 0.2390
....Saving...
Epoch 2
........mini-batch no. 12 loss: 1.0033
....train loss: 1.0022 :: EER 0.2843 :: Canc_EER 0.4725 :: D_sys 0.4943
....valid loss: 1.0070 :: EER 0.3723 :: Canc_EER 0.5000 :: D_sys 0.2673
....Saving...
Epoch 3
........mini-batch no. 12 loss: 1.0044
....train loss: 1.0014 :: EER 0.2703 :: Canc_EER 0.4828 :: D_sys 0.6079
....valid loss: 1.0039 :: EER 0.3298 :: Canc_EER 0.5000 :: D_sys 0.4158
....Saving...
Epoch 4
........mini-batch no. 12 loss: 1.0177
....train loss: 1.0010 :: EER 0.2293 :: Canc_EER 0.4754 :: D_sys 0.6700
....valid loss: 1.0041 :: EER 0.3340 :: Canc_EER 0.4681 :: D_sys 0.4386
Epoch 5
........mini-batch no. 12 loss: 1.0064
....train loss: 0.9995 :: EER 0.1872 :: Canc_EER 0.4697 :: D_sys 0.7646
....valid loss: 1.0001 :: 

Testing the Inception Resnet Model


In [43]:
!python iris_test_secure_inception_model.py

Testing model: Securetl_Iris_Inception_model_50
EER 0.2500 at threshold 0.0819 :: Canc_EER 0.4833 at threshold 0.0603 :: d_sys 0.6983
Privacy Leakage Rate 1.0000, Secrecy Leakage 1.3290


Summary



1.   Epochs = 5; EER 0.2000 at threshold 0.0674 :: Canc_EER 0.5000 at threshold 0.0519 :: d_sys 0.6529
Privacy Leakage Rate 1.0000, Secrecy Leakage 1.3290
2.   Epochs = 50; EER 0.2500 at threshold 0.0819 :: Canc_EER 0.4833 at threshold 0.0603 :: d_sys 0.6983
Privacy Leakage Rate 1.0000, Secrecy Leakage 1.3290



# Vision Transformer - Vit_b_16 model 

In [6]:
!python iris_train_securetl_attention_model.py

Training model: Securetl_Iris_Attention_model_16b_50
Downloading: "https://download.pytorch.org/models/vit_b_16-c867db91.pth" to /root/.cache/torch/hub/checkpoints/vit_b_16-c867db91.pth
100% 330M/330M [00:02<00:00, 159MB/s]
Epoch 1
........mini-batch no. 12 loss: 1.1253
....train loss: 1.0127 :: EER 0.2686 :: Canc_EER 0.4975 :: D_sys 0.5405
....valid loss: 1.0098 :: EER 0.2695 :: Canc_EER 0.5085 :: D_sys 0.5668
....Saving...
Epoch 2
........mini-batch no. 12 loss: 1.0873
....train loss: 1.0101 :: EER 0.2596 :: Canc_EER 0.4963 :: D_sys 0.5897
....valid loss: 1.0079 :: EER 0.2553 :: Canc_EER 0.5053 :: D_sys 0.5963
....Saving...
Epoch 3
........mini-batch no. 12 loss: 1.0439
....train loss: 1.0087 :: EER 0.2453 :: Canc_EER 0.4844 :: D_sys 0.6193
....valid loss: 1.0070 :: EER 0.2553 :: Canc_EER 0.5000 :: D_sys 0.5891
Epoch 4
........mini-batch no. 12 loss: 1.1371
....train loss: 1.0075 :: EER 0.2351 :: Canc_EER 0.4945 :: D_sys 0.6370
....valid loss: 1.0060 :: EER 0.2340 :: Canc_EER 0.5000 

In [8]:
!python iris_test_secure_attention_model.py

Testing model: Securetl_Iris_Attention_model_16b_50
EER 0.3000 at threshold 0.1385 :: Canc_EER 0.4833 at threshold 0.1078 :: d_sys 0.5210
Privacy Leakage Rate 1.0000, Secrecy Leakage 1.3290


# Result Analysis

Vit_b_32 model

In [44]:
!python results_analysis.py

My Method
EER 0.18333333333333332
FNMR@0.1%FMR 1.0
FNMR@1%FMR 0.5



In [47]:
!python results_analysis.py

My Method
EER 0.18333333333333332
FNMR@0.1%FMR 1.0
FNMR@1%FMR 0.5
FMR_C@EER 0.8333333333333334
d_sys 0.7767482447297911



Inception-resnet model

In [45]:
!python results_analysis.py

My Method
EER 0.25
FNMR@0.1%FMR 1.0
FNMR@1%FMR 0.8833333333333333



In [46]:
!python results_analysis.py

My Method
EER 0.25
FNMR@0.1%FMR 1.0
FNMR@1%FMR 0.8833333333333333
FMR_C@EER 0.7666666666666667
d_sys 0.6983252274290191



Vit_b_16 model

In [10]:
!python results_analysis.py

Vit_b_16
EER 0.3
FNMR@0.1%FMR 1.0
FNMR@1%FMR 0.8333333333333334
FMR_C@EER 0.7
d_sys 0.5210334966332928

